pennylane_lightning/src/Apply.cpp

```diff
@@ -106,3 +112,101 @@ void Pennylane::apply(
     }
 
 }
+
+void Pennylane::adjointJacobian(
+    StateVector& phi,
+    double* jac,
+    const vector<string>& observables,
+    const vector<vector<double> >& obsParams,
+    const vector<vector<unsigned int> >& obsWires,
+    const vector<string>& operations,
+    const vector<vector<double> >& opParams,
+    const vector<vector<unsigned int> >& opWires,
+    const vector<int>& trainableParams,
+    int paramNumber
+) {
+    vector<Pennylane::StateVector> lambdas;
+    size_t numObservables = observables.size();
+    size_t trainableParamNumber = trainableParams.size() - 1;
+
+    for (unsigned int i = 0; i < numObservables; i++) {
+        // copy |phi> and apply observables one at a time
+        CplxType* phiCopyArr = new CplxType[phi.length];
+        std::memcpy(phiCopyArr, phi.arr, sizeof(CplxType)*phi.length);
+        Pennylane::StateVector phiCopy(phiCopyArr, phi.length);
+
+        Pennylane::constructAndApplyOperation(
+            phiCopy,
+            observables[i],
+            obsWires[i],
+            obsParams[i],
+            false,
+            obsWires[i].size()
+        );
+        lambdas.push_back(phiCopy);
```

Comment:
So we know how large `lambdas` is going in.  Why not preallocate and fill? Is there a reason we need vector functionality?

Reply:
Ok. I think I see.  `lambdas` is a vector of statevectors? Why not just have an array with additional dimension?